objax/module.py

```diff
@@ -53,6 +53,54 @@ def __call__(self, *args, **kwargs):
         raise NotImplementedError
 
 
+class ForceArgs(Module):
+    """Forces override of arguments of given module."""
+
+    @staticmethod
+    def undo(module: Module):
+        """Undo ForceArgs on each submodule of the module.
+
+        Args:
+            module: module for which to undo ForceArgs.
+        """
+        if isinstance(module, ForceArgs):
+            raise ValueError('ForceArgs.undo can not be called on ForceArgs, call it on parent module instead.')
+        if isinstance(module, ModuleList):
+            for idx, v in enumerate(module):
+                if isinstance(v, Module):
+                    if isinstance(v, ForceArgs):
+                        v = v._wrapped
+                        module[idx] = v
+                    ForceArgs.undo(v)
+        else:
+            for k, v in module.__dict__.items():
+                if isinstance(v, Module):
+                    if isinstance(v, ForceArgs):
+                        v = v._wrapped
+                        module.__dict__[k] = v
+                    ForceArgs.undo(v)
```

Comment:
The code can be greatly simplified noting that:
1. `ForceArgs` called on a root of module tree, returns a new module.
2. Consequently calling undo on this root should also return a module, the original one.

Also I think `setattr` is the preferred way to write attributes in Python, so I also made that change.

```suggestion
    def undo(module: Module):
        """Undo ForceArgs on each submodule of the module.

        Args:
            module: module for which to undo ForceArgs.
        """
        if isinstance(module, ForceArgs):
            return ForceArgs.undo(module.__wrapped__)

        if isinstance(module, ModuleList):
            for idx, v in enumerate(module):
                if isinstance(v, Module):
                    module[idx] = ForceArgs.undo(v)
            return module

        for k, v in module.__dict__.items():
            if isinstance(v, Module):
                setattr(module, k, ForceArgs.undo(v))
        return module
```

Reply:
I think user should not be allowed to call `ForceArgs.undo` on instance of `ForceArgs` because it may lead to surprising behavior from user point of view:

```python
# override args of one operation in the model
m = SomeModel()
m.bn1 = ForceArgs(m.bn1, training=False)
y = m(x, training=True)  # argument of m.bn1 is overriden

ForceArgs.undo(m)  # reset all overrides

# override args of entire model
pred_m = ForceArgs(m, training=False)
y_pred = pred_m(x)  # training is overriden for entire model

# if ForceArgs.undo can be called on instance of ForceArgs following line will do nothing
# however user may expect that arguments override will be undone
ForceArgs.undo(pred_m)
```

There is a separate question whether `ForceArgs.undo(m)` should return `m`.
I think it's ok to add, but not sure how useful. I just allow users to write code like `y = ForceArgs.undo(m)(x)` which might be actually somewhat confusing

Reply:
1. One can call `ForceArgs` on the root, therefore calling `ForceArgs.undo` should be doable on the root too.
2. Returning a module results in simpler implementation too.